db/hybrid_logical_vector.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023-Present Couchbase, Inc.
+//
+// Use of this software is governed by the Business Source License included
+// in the file licenses/BSL-Couchbase.txt.  As of the Change Date specified
+// in that file, in accordance with the Business Source License, use of this
+// software will be governed by the Apache License, Version 2.0, included in
+// the file licenses/APL2.txt.
+
+package db
+
+import (
+	"encoding/json"
+
+	"github.com/couchbase/sync_gateway/base"
+)
+
+type HybridLogicalVector struct {
+	CurrentVersionCAS uint64            // current version cas (or cvCAS) stores the current CAS at the time of replication
+	SourceID          string            // source bucket uuid of where this entry originated from
+	Version           uint64            // current cas of the current version on the version vector
+	MergeVersions     map[string]uint64 // map of merge versions for fast efficient lookup
+	PreviousVersions  map[string]uint64 // map of previous versions for fast efficient lookup
+}
+
+type PersistedHybridLogicalVector struct {
+	CurrentVersionCAS string            `json:"cvCas"` // current version cas (or cvCAS) stores the current CAS at the time of replication
+	SourceID          string            `json:"src"`   // source bucket uuid of where this entry originated from
+	Version           string            `json:"vrs"`   // current cas of the current version on the version vector
+	MergeVersions     map[string]string `json:"mv"`    // map of merge versions for fast efficient lookup
+	PreviousVersions  map[string]string `json:"pv"`
+}
+
+type PersistedVersionVector struct {
+	PersistedHybridLogicalVector `json:"_vv"`
+}
+
+func NewHybridLogicalVector() HybridLogicalVector {
+	return HybridLogicalVector{
+		MergeVersions:    make(map[string]uint64),
+		PreviousVersions: make(map[string]uint64),
+	}
+}
+
+func (hlv *HybridLogicalVector) MarshalJSON() ([]byte, error) {
+	persistedHLV := PersistedVersionVector{}
+	cvCasByteArray := base.Uint64CASToLittleEndianHex(hlv.CurrentVersionCAS)
+	var persistedSRC string
+	// check if it is a server mutation, if not we need to convert to base64 encoding and add the mobile mutation prefix
+	if hlv.SourceID[0:2] != "s_" {
+		encodedValue, err := base.HexToBase64(hlv.SourceID)
```

Comment:
I don't think we want to make any assumption about the format of non-server SourceIDs (i.e. that they are hex values).  Replicating peers *should* be defining a storage-optimized sourceID value, but I think it's going to be that peer's responsibility to do so, and SG shouldn't do any additional formatting.

Reply:
More generally, I think it makes sense for the in-memory storage of the sourceIDs to match the format used during persistence, to simplify triaging issues where we're comparing log output to persisted HLVs.

I understand this makes it more work to get from a sourceID back to a bucket UUID, but that seems like the less common comparison of the two.